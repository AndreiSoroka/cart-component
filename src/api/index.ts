export default function api(input: RequestInfo | URL, init?: RequestInit) {
  // add some config if you need
  return fetch(input, init).then((response) => response.json());
}
