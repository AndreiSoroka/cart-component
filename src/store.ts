import { configureStore } from '@reduxjs/toolkit'
import CartSlice from "./features/ShopCart/stores/cart/cart.slice";
import ShopsSlice from "./features/ShopCart/stores/shops/shops.slice";

const store = configureStore({
  reducer: {
    cart: CartSlice.reducer,
    shops: ShopsSlice.reducer
  }
})

export default store

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
