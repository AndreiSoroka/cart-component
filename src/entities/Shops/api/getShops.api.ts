import { createAsyncThunk } from "@reduxjs/toolkit";
import api from "@/shared/api";
import isShopItems from "@/entities/Shops/guards/isShopItems.guard";
import sortShopsFromApi from "@/entities/Shops/lib/helpers/sortShopsFromApi";

export const getShops = createAsyncThunk("shops/getShops", async () => {
  const list = await api("/api/shops.json");
  if (!isShopItems(list)) {
    throw new Error("Something wrong");
  }
  // todo id is u
  return sortShopsFromApi(list);
});
