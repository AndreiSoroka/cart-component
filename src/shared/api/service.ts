import { Api } from "@/shared/api/api.generated";
export default new Api();
