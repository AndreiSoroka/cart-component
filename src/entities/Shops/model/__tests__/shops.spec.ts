import { describe, expect, it, beforeEach } from "@jest/globals";
import { getShops } from "@/entities/Shops/api/getShops.api";
import { setupStore } from "@/store";
import fetchMock from "jest-fetch-mock";

const store = setupStore();
const MockShop1 = { id: "1", name: "Shop 1", sortOrder: 1 };
const MockShop2 = { id: "2", name: "Shop 2", sortOrder: 2 };
const MockShop3 = { id: "3", name: "Shop 3", sortOrder: 3 };
const MockShops = [MockShop2, MockShop1, MockShop3];

fetchMock.enableMocks();

describe("Shops reducer", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });

  it("should handle initial state", () => {
    const { list, status, error } = store.getState().shops;
    expect(list).toMatchSnapshot();
    expect(status).toBe("idle");
    expect(error).toBe("");
  });

  it("should handle getShops", async () => {
    fetchMock.mockResponseOnce(JSON.stringify(MockShops));

    await store.dispatch(getShops());
    const { list, status, error } = store.getState().shops;
    expect(list).toMatchSnapshot();
    expect(status).toBe("success");
    expect(error).toBe("");
  });

  it("should handle getShops normal error", async () => {
    fetchMock.mockResponseOnce(
      JSON.stringify({
        errorId: "errorId",
        errorMessage: "errorMessage 123",
      }),
      { status: 500 }
    );

    await store.dispatch(getShops());
    const { status, error } = store.getState().shops;
    expect(status).toBe("error");
    expect(error).toBe("errorMessage 123");
  });

  it("should handle getShops error", async () => {
    fetchMock.mockResponseOnce("", { status: 500 });

    await store.dispatch(getShops());
    const { error, status } = store.getState().shops;
    expect(error).toContain("Something wrong");
    expect(status).toBe("error");
  });

  it("should handle error when fetch is rejected without a message", async () => {
    const errorWithoutMessage = new Error();
    fetchMock.mockRejectOnce(errorWithoutMessage);

    await store.dispatch(getShops());
    const { error, status } = store.getState().shops;
    expect(error).toBe("Something wrong");
    expect(status).toBe("error");
  });
});
