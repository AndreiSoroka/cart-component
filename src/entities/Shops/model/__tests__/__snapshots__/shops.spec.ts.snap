// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shops reducer should handle getShops 1`] = `
{
  "entities": {
    "1": {
      "id": "1",
      "name": "Shop 1",
      "sortOrder": 1,
    },
    "2": {
      "id": "2",
      "name": "Shop 2",
      "sortOrder": 2,
    },
    "3": {
      "id": "3",
      "name": "Shop 3",
      "sortOrder": 3,
    },
  },
  "ids": [
    "1",
    "2",
    "3",
  ],
}
`;

exports[`Shops reducer should handle initial state 1`] = `
{
  "entities": {},
  "ids": [],
}
`;
